[Albert Ariel Widiaatmaja] iP

albertarielw opened the pull request:
### LishBot v6.9

> "Imagine Duke, but way better" - some random dude in my class one day

LishBot offers what Duke offers with some extra functionalities:

- More types of tasks: todo, event, deadline
- Cool date format

How to get it?
1. [download here](https://github.com/albertarielw/ip/releases)
2. ~wait what there is only one step~ :frowning:

How to run it?

run the following command on the same directory as your LishBot

> java -jar LishBot.jar

Not impressed yet, just look at this cool code:

```
public class DukeException extends Exception {
    DukeException(String errorMessage) {
        super(errorMessage);
    }

    @Override
    public String toString() {
        return super.getMessage();
    }
}
```

Commit:
Add Gradle support

albertarielw committed:
Level-1

albertarielw committed:
Level-2

albertarielw committed:
Level-3

albertarielw committed:
Level-4

albertarielw committed:
A-TextUiTesting

albertarielw committed:
Level-5

albertarielw committed:
Level-6

albertarielw committed:
Add more personality

albertarielw committed:
branch-level-7 add save data

albertarielw committed:
branch-Level-8 Add Local Date

albertarielw committed:
Merge tag 'branch-Level-7'

albertarielw committed:
Resolved merge conflict branch-Level-7 and branch-Level-8

albertarielw committed:
more OOP (almost done)

albertarielw committed:
A-MoreOOP

albertarielw committed:
A-Packages

albertarielw committed:
Merge branch 'add-gradle-support' of https://github.com/nus-cs2103-AY2223S1/ip

* 'add-gradle-support' of https://github.com/nus-cs2103-AY2223S1/ip:
  Add Gradle support

albertarielw committed:
A-Gradle add gradle to accelerate build

albertarielw committed:
A-JUnit add JUnit test for Parser and TaskList and run it using gradle

albertarielw committed:
A-Jar build and test jar file

albertarielw committed:
branch-Level-9

albertarielw committed:
branch-A-CodingStandard rewrite lines to follow coding standard

albertarielw committed:
branch-A-JavaDoc add JavaDoc comments

albertarielw committed:
Merge tag 'branch-Level-9'

albertarielw committed:
Merge tag 'branch-A-JavaDoc'

# Conflicts:
#	src/main/java/duke/TaskList.java
#	src/main/java/duke/UI.java

albertarielw committed:
Merge tag 'branch-A-CodingStandard'

albertarielw committed:
Merge JavaDoc, CodingStandard and Level-9 to master

albertarielw committed:
Make code neater

deeyonn commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,85 @@
+        } catch (DukeException e) {
+            UI.printResponse(e.toString());
+        }
+    }
```
I see that you are using switch statements! 😄 
Consider the following notes:
- Avoid long methods (keep methods shorter than a screen)
- Avoid deep nesting (indentation level should be reasonably low)

deeyonn commented on src/main/java/duke/Event.java:
```diff
@@ -0,0 +1,32 @@
+    @Override
+    public String toString() {
+        return "[" + getTaskTypeIcon() + "]" + super.toString() + " (" + timeQualifier + ": " + timeDescription.format(DateTimeFormatter.ofPattern("MMM d yyyy")) + ")\n";
+    }
```
Its nice to see method overriding being used! 😄 
Consider the following notes when structuring statements:
- Avoid complicated expressions (consider breaking them up into smaller pieces)

deeyonn commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,86 @@
+        } catch (IOException e) {
+            throw new DukeException("IOException error has occurred :(\n");
+        }
+    }
```
There seems to be alot going on here 😢 Could it be possible to extract some logic to flatten out the code?
- Avoid long methods (keep methods shorter than a screen)
- Avoid deep nesting (indentation level should be reasonably low)

deeyonn commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,127 @@
+
+        for (int i = 0; i < taskList.size(); ++i) {
+            Task task = taskList.get(i);
+            if (task.getDescription().toLowerCase().matches(("(.*)" + description + "(.*)").toLowerCase())) {
```
You can consider extracting this statement into a boolean for easier readability.
- Avoid complicated expressions (consider breaking them up into smaller pieces)
I noticed several instances of this during the review
```suggestion
boolean isCondition = task.getDescription().toLowerCase().matches(("(.*)" + description + "(.*)").toLowerCase());
if (isCondition) {
```

deeyonn commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,127 @@
+     */
+    public void markTaskAsDone(int index) throws DukeException {
+        if (index < 0) {
+            throw new DukeException("I cannot mark this task because your index is invalid");
+        } else if (index >= taskList.size()) {
+            throw new DukeException("I cannot mark this task because task does not exist");
```
Consider extracting out frequently used statements as constants.
```suggestion
public static final String ERROR_INVALID_INDEX = "I cannot mark this task because your index is invalid";
public static final String ERROR_DOES_NOT_EXIST = "I cannot mark this task because task does not exist";
    throw new DukeException(ERROR_INVALID_INDEX);
} else if (index >= taskList.size()) {
    throw new DukeException(ERROR_DOES_NOT_EXIST);
```

deeyonn commented on src/main/java/duke/UI.java:
```diff
@@ -0,0 +1,31 @@
+package duke;
+
+public class UI {
+    public static String breakLine = "____________________________________________________________\n";
+    public static String greeting = "Hello, I'm LishBot v6.9!\n" + "How may I help you today?\n";
+    public static String taskListOpening = "Finding your task list...\n" + "Found it! Here are what you have to do:\n";
+    public static String noListFound = "Congrats! You have finished all your task!\n";
+    public static String findingRelatedTask = "Let me find tasks that match your description...\n";
+    public static String noRelatedTaskFound = "I cannot find any tasks that match that description :(\n";
+    public static String goodBye = "Glad to be of help! See you later~\n";
+    public static String markAsDone = "Great! You have finished a task. I will mark this as done\n";
+    public static String markAsNotDone = "Okay. I will mark this task as not done yet\n";
+    public static String removeTask = "Roger that! I will remove this task\n";
+    public static String numberOfTaskLeft = "Now, the number of tasks you have is ";
+    public static String addTask = "Got it! I will add that task now\n";
+    public static String fileNotFound = "I cannot find the data file \n";
+    public static String createFile = "But don't worry I will create one for you :)\n";
```
Do consider that some of these can have the final flag and have their names fit the coding standard for constants.
```suggestion
    public static final String MSG_LINE_BREAK = "____________________________________________________________\n";
    public static final String MSG_GREETING = "Hello, I'm LishBot v6.9!\n" + "How may I help you today?\n";
    public static final String MSG_TASK_LIST = "Finding your task list...\n" + "Found it! Here are what you have to do:\n";
    public static final String MSG_CONGRATS = "Congrats! You have finished all your task!\n";
    public static final String MSG_FIND_MATCH = "Let me find tasks that match your description...\n";
    public static final String MSG_GOODBYE = "Glad to be of help! See you later~\n";
    public static final String MSG_MARK_DONE = "Great! You have finished a task. I will mark this as done\n";
    public static final String MSG_MARK_NOT_DONE = "Okay. I will mark this task as not done yet\n";
    public static final String MSG_REMOVE_TASK = "Roger that! I will remove this task\n";
    public static final String MSG_NUMBER_OF_TASKS = "Now, the number of tasks you have is ";
    public static final String MSG_ADD_TASK = "Got it! I will add that task now\n";
    public static final String MSG_FILE_CREATED = "But don't worry I will create one for you :)\n";
        
    public static final String ERR_CANNOT_FIND = "I cannot find any tasks that match that description :(\n";
    public static final String ERR_DATA_NOT_FOUND = "I cannot find the data file \n";
```

deeyonn reviewed (commented):
Excellent work!! 😄 Some general items to note on coding quality:
- Avoid long methods (keep methods shorter than a screen)
- Avoid deep nesting (indentation level should be reasonably low)
- Avoid complicated expressions (consider breaking them up into smaller pieces)
- Avoid magic numbers (consider using constants)
- Practice single level of abstraction
- Make the execution path when everything goes well prominent

albertarielw committed:
fix typos

albertarielw committed:
add JavaFX environment

albertarielw committed:
handle user input cases

albertarielw committed:
handle greeting and goodbye

albertarielw committed:
add FXML

albertarielw committed:
implement FXML

albertarielw committed:
Level-10 finish GUI, add greeting

albertarielw committed:
Merge branch 'branch-Level-10'

mlzt2000 commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,102 @@
+                try {
+                    String[] commands = input.split(" ");
+
+                    if (commands.length < 2) {
```
Do you think using a nested switch statements instead of many if-else statements would be less confusing in this case?

mlzt2000 commented on src/main/java/duke/Event.java:
```diff
@@ -0,0 +1,32 @@
+
+    @Override
+    public String toWrite() {
+        return getTaskTypeIcon() + "," + (super.isDone ? "1," : "0,") + super.description + "," + timeQualifier + "," + timeDescriptionInput + "\n";
```
Since you are using quite a long String concatenation here, would String.format() make things clearer?

albertarielw committed:
update jar

albertarielw committed:
add Assertions

albertarielw committed:
Merge pull request #1 from albertarielw/branch-A-Assertions

add Assertions

albertarielw committed:
Merge tag 'branch-A-Assertions'

albertarielw committed:
Merge branch 'master' of https://github.com/albertarielw/ip

albertarielw committed:
refactor to improve code quality

albertarielw committed:
Merge branch 'branch-A-CodeQuality'

albertarielw committed:
Fix Typos

albertarielw committed:
Merge pull request #2 from albertarielw/branch-A-CodeQuality

Improve Code Quality

albertarielw committed:
D-Trivia: add memorize functionality to chatbot

This functionality allows users to personalize their chatbot and chatbot's responses are not limited to what it is initially programmed.
In fact, with this functionality, the chatbot can have significantly more distinct responses to different instructions.

Use memorize functionality by using the following instruction format: "teach [qns]/[ans]".
The bot will respond with ans whenver it receives qns
The bot is able to store qns and ans in external storage. Hence upon closing the bot, the bot still remembers all past trivia.

albertarielw committed:
fix typos

albertarielw committed:
Set up CI

albertarielw committed:
add exit functionality

albertarielw committed:
update jar

albertarielw committed:
Release jar

albertarielw committed:
add Ui screenshot

albertarielw committed:
Update README.md

albertarielw committed:
Update README.md

albertarielw committed:
Update README.md

albertarielw committed:
Update README.md

albertarielw committed:
Update README.md

albertarielw committed:
Update README.md

albertarielw committed:
Update README.md

albertarielw committed:
Update README.md

albertarielw committed:
Update README.md

albertarielw committed:
Update README.md

albertarielw committed:
Update README.md

albertarielw committed:
Update README.md

albertarielw committed:
add User Guide

albertarielw committed:
improve UI

albertarielw committed:
improve chat bubble UI

albertarielw committed:
fix anchors

albertarielw committed:
update icons

albertarielw committed:
fix packages

albertarielw committed:
Add more JavaDocs comments

albertarielw committed:
Update screenshot

albertarielw committed:
Update README.md

albertarielw committed:
Update README.md

albertarielw committed:
Adds more JavaDocs comments

albertarielw committed:
Merge branch 'master' of https://github.com/albertarielw/ip

albertarielw committed:
Update README.md

albertarielw committed:
Update README.md

albertarielw committed:
Update README.md

albertarielw committed:
Update README.md

albertarielw committed:
Adds user guide theme

albertarielw committed:
customize theme

albertarielw committed:
Update README.md

albertarielw committed:
Update README.md

albertarielw committed:
Update README.md

albertarielw committed:
fix package naming issue
